### PR TITLE
ci: draft a playbook

### PR DESCRIPTION
**Pull request by jhchabran**
Reviewing: please do not pay attention to the form, it's the content that matters here! 

- Do you think you'd be confident by having this playbook to be a build sheriff? 
- Is this detailed enough given the target audience?
- Are the steps easy to follow?

### CODE REVIEW

File: content/engineering/incidents/playbooks/ci.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+In order to handle problems with the CI:
+
+- Have access to the `sourcegraph-ci` _project_ on Google Cloud Platform.
+  - See [#it-tech-ops](https://sourcegraph.slack.com/archives/C01CSS3TC75)
+- Have the CLI `gcloud` tool installed and have authenticated yourself.
+  - See [Gain access to the cluster](../../engineering/deployments/debugging/tutorial/index.md)
+- Set the current Kubernetes context to the `default-buildkite` _cluster_ in the `sourcegraph-ci` _project_:
+  - `gcloud config set project sourcegraph-ci`
+  - `gcloud container clusters get-credentials default-buildkite --zone us-central1-c`
```
Comment (bobheadxi) [inline content/engineering/incidents/playbooks/ci.md]: I think this should be part of `Gain access to the cluster` (if it isn't already)

File: content/engineering/incidents/playbooks/ci.md [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+  - The `main` branch runs additional checks compared to Pull Requests builds. So it's possible that one of those checks failed.
+  - The `main` branch have changes that weren't in the Pull Request branch and those changes are causing a failure.
+  - The `main` branch is failing due to a previous build.
+
+#### Actions
+
+1. Check your build on [Buildkite](https://buildkite.com/sourcegraph/sourcegraph/builds?branch=main)
+
+- Find its link directly in the [#buildkite-main](https://sourcegraph.slack.com/archives/C02FLQDD3TQ) channel.
+- :bulb: Or run `sg ci status` in your shell, with the same branch checked out.
```
Comment (bobheadxi) [inline content/engineering/incidents/playbooks/ci.md]: There's a few mentions of "the same branch" here, maybe we should just keep it as `main` for simplicity, or even give `--branch=main`?

File: content/engineering/incidents/playbooks/ci.md [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+2. Search for the failing steps, and browse the logs (:bulb: run `sg ci logs` in your shell, with the same branch checked out) .
+
+- Look for a failure explanation: it can be a test that failed or a command that return a non zero exit code.
+
+3. Check the previous builds on the `main` branch on [Buildkite](https://buildkite.com/sourcegraph/sourcegraph/builds?branch=main)
```
Comment (bobheadxi) [inline content/engineering/incidents/playbooks/ci.md:47]: We should just refer to Loki here, perhaps an example query to demonstrate searching for a message

File: content/engineering/incidents/playbooks/ci.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+1. Are they failing with the same exact error?
+
+   - Yes: see the [Builds are failing in the `main` branch with the same error]()
+   - No: see next point.
+
+1. Is that a real failure or a flake?
+1. Restart that step. Maybe it will fail again, but if it doesn't it'll save you time.
```
Comment (bobheadxi) [inline content/engineering/incidents/playbooks/ci.md]: Are these two steps related? (might be formatting)

File: content/engineering/incidents/playbooks/ci.md [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+- Impact: no commits are being deployed on DogFood and `sourcegraph.com` until the problem is resolved. Cutting a release is impossible.
+- Possible causes:
+  - A previous Pull Request introduced a change that causes a test to fail.
+  - A previous Pull Request introduced a change that modified state in an unexpected way and broke the CI.
+  - An external dependency is not available anymore and is causing builds to fail.
+  - Some rate limiting API is throttling us and causing builds to fail.
+
+#### Actions
+
+1. Identify the error in common with the recent builds on [Buildkite](https://buildkite.com/sourcegraph/sourcegraph/builds?branch=main).
```
Comment (bobheadxi) [inline content/engineering/incidents/playbooks/ci.md:86]: Also search Loki

File: content/engineering/incidents/playbooks/ci.md [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+1. Is this an external failure or an internal one?
+
+- :bulb: External failures are about downloading a dependency like a package in a script or a in a Dockerfile. Often they'll manifest in the form of an HTTP error.
+- :bulb: If unsure, ask for help on [#dev-chat](https://sourcegraph.slack.com/archives/C07KZF47K).
+- Yes, it's an external failure:
+  1. SSH into an agent to reproduce the problem:
+  1. Launch `k9s` and select an agent, they are named like `buildkite-agent-xxxxxxxxxx-yyyyy`
+  1. Press `s` to SSH on it.
+  1. You will be prompted with which pod to SSH onto, select the agent itself.
```
Comment (bobheadxi) [inline content/engineering/incidents/playbooks/ci.md]: This could be a separate playbook entry I think (direct access to agent)

File: content/engineering/incidents/playbooks/ci.md [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+#### Actions
+
+1. Escalate by creating an incident (`/incident` on Slack).
+1. Get some help.
+1. Downscale the agents to be able to observe exactly what's going on.
+1. Update the `autoscaler` manifest to run a single agent
+1. `cd sourcegraph/infrastructure`
+1. Edit the [manifest](https://sourcegraph.com/github.com/sourcegraph/infrastructure/-/blob/buildkite/kubernetes/buildkite-autoscaler/buildkite-autoscaler.Deployment.yaml?L32-35) to have 1 on both max an min agent count.
+1. `cd buildkite/kubernetes/buildkite-autoscaler`
+1. Run `kubectl apply -n buildkite -f buildkite-autoscaler.Deployment.yaml`
```
Comment (bobheadxi) [inline content/engineering/incidents/playbooks/ci.md]: ALso think this should be a standalone guide for making change

**Review by bobheadxi — COMMENTED** [review]
quick skim - this is a lot of content! will review again later 😁  thanks for putting this together!

File: content/engineering/incidents/playbooks/ci.md (hunk 1 above)
Comment (jhchabran) [reply content/engineering/incidents/playbooks/ci.md]: That page does mention these instructions, just not for the CI cluster. I thought it would be useful to have this here, it makes this self-contained in some way. 

Still, I agree with you that it should be present on that page, I've added some content in there.

**Comment by jhchabran** [comment]
This is now up for review (it's a bit hard to review yourself on such content where every line looks the same, I'm sorry in advance if you find obvious mistakes).

**Comment by jhchabran** [comment]
@daxmc99, @davejrt and @ggilmore, I pinged you on this for your eventual feedback on a higher level regarding those steps, please don't bother yourself with any grammar mistake or templating issue 🙏

File: content/engineering/deployments/debugging/tutorial.md [hunk 8]
```diff
@@ -37,7 +37,45 @@ dogfood-server    Active   77d
 ...
 ```
 
-Now, change to the `dogfood-k8s` namespace via `kubectl config set-context --current --namespace=dogfood-k8s` to work with workloads in that namespace.
+---
```
Comment (bobheadxi) [inline content/engineering/deployments/debugging/tutorial.md]: Let's split this into per-deployment headings to make it easier to reference

File: content/engineering/deployments/debugging/tutorial.md [hunk 9]
```diff
@@ -37,7 +37,45 @@ dogfood-server    Active   77d
...
+cluster-ci-5967   Active   28d
+default           Active   407d
+kube-node-lease   Active   407d
+kube-public       Active   407d
+kube-system       Active   407d
+...
+```
+
+Now, change to the `buildkite` namespace via `kubectl config set-context --current --namespace=buildkite` to work with workloads in that namespace.
 Note: You do not need to switch the namespace to access workloads in other namespaces, you can use `-n ${NAMESPACE}` with your command to access other namespaces.
```
Comment (bobheadxi) [inline content/engineering/deployments/debugging/tutorial.md:81]:
```suggestion
> NOTE: You do not need to switch the namespace to access workloads in other namespaces, you can still use `-n ${NAMESPACE}` with your command to access other namespaces.
```

File: content/engineering/deployments/debugging/tutorial.md (hunk 8 above)
Comment (bobheadxi) [reply content/engineering/deployments/debugging/tutorial.md]: Also up to you how you want to combine these two, but there's another page https://handbook.sourcegraph.com/engineering/deployments/instances with partial instructions on accessing clusters

File: content/engineering/incidents/playbooks/ci.md [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+The CI is what enables us to feel confident when delivering our changes to our users, and is one of the key components enabling Sourcegraph to deliver quality software. While the DevX team is in charge of managing the CI as a tool, it is essential for every engineer to be able to unblock themselves if there is a problem in order be autonomous.
+
+This page lists common failures scenarios and provide a step by step guide to get the CI back in an operational state.
+
+## Scenarios
+
+### Build has failed on the `main` branch
+
+- Gravity: _minor_
+- Impact: that commit won't be deployed on DogFood and `sourcegraph.com` until an ulterior build passes.
```
Comment (bobheadxi) [inline content/engineering/incidents/playbooks/ci.md:33]:
```suggestion
- Impact: that commit won't be deployed on `k8s.sgdev.org` and `sourcegraph.com` until a later build passes.
```

File: content/engineering/incidents/playbooks/ci.md [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+- Possible causes:
+  - The `main` branch runs additional checks compared to Pull Requests builds. So it's possible that one of those checks failed.
```
Comment (bobheadxi) [inline content/engineering/incidents/playbooks/ci.md:35]: Should note + link to stuff about how our pipeline is generated based on a variety of factors

File: content/engineering/incidents/playbooks/ci.md [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+
+- Gravity: _minor_
+- Impact: May fail ongoing builds, but that's fine.
+- Possible causes:
+  - Manual restart by an engineer.
+  - Newer version of the agents needs to be deployed.
+
+#### Actions
+
+1. Open `k9s` to observe the currently running agents.
```
Comment (bobheadxi) [inline content/engineering/incidents/playbooks/ci.md]: Are we recommending `k9s` for everyone? (I personally have never tried it before)

Comment (jhchabran) [reply content/engineering/incidents/playbooks/ci.md]: I don't think that suggests that this is the only way to go.

Either you know your way with `kubectl` and you'll grasp `k9s` in like two minutes or you don't, in which case, `k9s` will make it much simpler to interact with the cluster.

Comment (daxmc99) [reply content/engineering/incidents/playbooks/ci.md]: I would suggest to use `kubectl get pods -w`  
Let's keep our playbook docs fairly generic 
```suggestion
1. Use`kubectl get pods -w` to observe the currently running agents (`k9s` works here too).
```

File: content/engineering/incidents/playbooks/ci.md [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+- Gravity: none
+- Impact: none (unless a destructive action is performed)
+- Possible cause:
+  - Need to investigate a problem and suspect the agent is at fault
+
+#### Actions
+
+1. SSH into an agent to reproduce the problem:
+1. Launch `k9s` and select an agent, they are named like `buildkite-agent-xxxxxxxxxx-yyyyy`
+1. Press `s` to SSH on it.
```
Comment (daxmc99) [inline content/engineering/incidents/playbooks/ci.md]: Same as above, I would suggest `kubectl exec -it` here
```suggestion
1. Use `kubectl exec -it buildkite-agent-xxxx-yyyy` to open a shell 
```
Also, `kubectl exec` is not the same as SSH 😉 https://goteleport.com/blog/ssh-vs-kubectl/

File: content/engineering/incidents/playbooks/ci.md [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+1. You will be prompted with which pod to SSH onto, select the agent itself.
```
Comment (daxmc99) [inline content/engineering/incidents/playbooks/ci.md]:
```suggestion
1. You will be prompted with which pod to `exec` onto, select the agent itself.
```

**Review by daxmc99 — COMMENTED** [review]
Looks good! 
All my comments are nits, this looks like a great playbook
